File: app/src/main/java/com/baijum/ukufretboard/viewmodel/MetronomeViewModel.kt
package com.baijum.ukufretboard.viewmodel

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.baijum.ukufretboard.audio.ClickSoundPlayer
import com.baijum.ukufretboard.audio.MetronomeEngine
import com.baijum.ukufretboard.data.BeatType
import com.baijum.ukufretboard.domain.MetronomeStateLogic
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

/**
 * ViewModel for the standalone metronome screen.
 *
 * Manages BPM, time signature, subdivision, accent pattern, and playback state.
 * Plays audible click sounds on each beat via [ClickSoundPlayer].
 */
class MetronomeViewModel(
    application: Application,
) : AndroidViewModel(application) {
    private val engine = MetronomeEngine()
    private val clickPlayer = ClickSoundPlayer(application)

    private val _bpm = MutableStateFlow(100)
    val bpm: StateFlow<Int> = _bpm.asStateFlow()

    private val _beatsPerMeasure = MutableStateFlow(4)
    val beatsPerMeasure: StateFlow<Int> = _beatsPerMeasure.asStateFlow()

    private val _subdivision = MutableStateFlow(1)
    val subdivision: StateFlow<Int> = _subdivision.asStateFlow()

    private val _accentPattern = MutableStateFlow(MetronomeStateLogic.defaultAccentPattern(4))
    val accentPattern: StateFlow<List<BeatType>> = _accentPattern.asStateFlow()

    private val _isPlaying = MutableStateFlow(false)
    val isPlaying: StateFlow<Boolean> = _isPlaying.asStateFlow()

    private val _currentBeat = MutableStateFlow(-1)
    val currentBeat: StateFlow<Int> = _currentBeat.asStateFlow()

    private val _currentSubBeat = MutableStateFlow(0)
    val currentSubBeat: StateFlow<Int> = _currentSubBeat.asStateFlow()

    private val _measureCount = MutableStateFlow(0)
    val measureCount: StateFlow<Int> = _measureCount.asStateFlow()

    private val _isCompound = MutableStateFlow(false)
    val isCompound: StateFlow<Boolean> = _isCompound.asStateFlow()

    private val _timeSignatureLabel = MutableStateFlow("4/4")
    val timeSignatureLabel: StateFlow<String> = _timeSignatureLabel.asStateFlow()

    fun setBpm(value: Int) {
        _bpm.value = MetronomeStateLogic.clampBpm(value)
        if (_isPlaying.value) restartEngine()
    }

    fun setTimeSignature(label: String) {
        _timeSignatureLabel.value = label
        val result = MetronomeStateLogic.parseTimeSignature(label)
        _isCompound.value = result.isCompound
        _beatsPerMeasure.value = result.beatsPerMeasure
        _subdivision.value = result.subdivision
        _accentPattern.value = MetronomeStateLogic.defaultAccentPattern(result.beatsPerMeasure)
        stop()
    }

    fun setBeatsPerMeasure(value: Int) {
        _beatsPerMeasure.value = value.coerceIn(2, 7)
        _accentPattern.value = MetronomeStateLogic.defaultAccentPattern(value)
        _isCompound.value = false
        _timeSignatureLabel.value = "$value/4"
        stop()
    }

    fun setSubdivision(value: Int) {
        if (_isCompound.value) return
        _subdivision.value = MetronomeStateLogic.clampSubdivision(value)
        stop()
    }

    fun toggleBeatType(beatIndex: Int) {
        val pattern = _accentPattern.value.toMutableList()
        if (beatIndex in pattern.indices) {
            pattern[beatIndex] = MetronomeStateLogic.cycleBeatType(pattern[beatIndex])
            _accentPattern.value = pattern
        }
    }

    fun togglePlayback() {
        if (_isPlaying.value) stop() else start()
    }

    private fun start() {
        _isPlaying.value = true
        _measureCount.value = 0
        _currentBeat.value = -1
        startEngine()
    }

    fun stop() {
        engine.stop()
        _isPlaying.value = false
        _currentBeat.value = -1
        _currentSubBeat.value = 0
    }

    private fun startEngine() {
        engine.startMetronome(
            scope = viewModelScope,
            bpm = _bpm.value,
            beatsPerMeasure = _beatsPerMeasure.value,
            subdivision = _subdivision.value,
            accentPattern = _accentPattern.value,
            onTick = { beat, subBeat, type ->
                _currentBeat.value = beat
                _currentSubBeat.value = subBeat
                when {
                    type == BeatType.MUTE -> {}

                    subBeat > 0 -> {
                        clickPlayer.playSubdivision()
                    }

                    type == BeatType.ACCENT -> {
                        clickPlayer.playAccent()
                    }

                    else -> {
                        clickPlayer.playNormal()
                    }
                }
            },
            onMeasure = { count ->
                _measureCount.value = count
            },
        )
    }

    private fun restartEngine() {
        engine.stop()
        startEngine()
    }

    override fun onCleared() {
        engine.stop()
        clickPlayer.release()
        super.onCleared()
    }
}

File: iosApp/UkuleleCompanion/ViewModels/MetronomeViewModel.swift
import Foundation
import Combine
import shared

enum BeatType: String {
    case accent, normal, mute
}

@MainActor
final class MetronomeViewModel: ObservableObject {
    @Published var bpm: Double = 100
    @Published var beatsPerMeasure: Int = 4
    @Published var subdivision: Int = 1
    @Published var accentPattern: [BeatType] = [.accent, .normal, .normal, .normal]
    @Published var isPlaying: Bool = false
    @Published var currentBeat: Int = -1
    @Published var currentSubBeat: Int = 0
    @Published var measureCount: Int = 0
    @Published var isCompound: Bool = false
    @Published var timeSignatureLabel: String = "4/4"

    private let tonePlayer = TonePlayer()
    private var timer: AnyCancellable?
    private var tapTimestamps: [Date] = []

    func setBpm(_ value: Double) {
        bpm = Double(MetronomeStateLogic.shared.clampBpm(value: Int32(value)))
        if isPlaying { restartTimer() }
    }

    func setTimeSignature(_ label: String) {
        timeSignatureLabel = label
        let result = MetronomeStateLogic.shared.parseTimeSignature(label: label)
        isCompound = result.isCompound
        beatsPerMeasure = Int(result.beatsPerMeasure)
        subdivision = Int(result.subdivision)
        accentPattern = Self.convertAccentPattern(
            MetronomeStateLogic.shared.defaultAccentPattern(beatsPerMeasure: result.beatsPerMeasure)
        )
        stop()
    }

    func setSubdivision(_ value: Int) {
        guard !isCompound else { return }
        subdivision = Int(MetronomeStateLogic.shared.clampSubdivision(value: Int32(value)))
        if isPlaying { restartTimer() }
    }

    func toggleBeatType(_ index: Int) {
        guard index >= 0, index < accentPattern.count else { return }
        accentPattern[index] = Self.convertBeatType(
            MetronomeStateLogic.shared.cycleBeatType(
                current: Self.toKmpBeatType(accentPattern[index])
            )
        )
    }

    func togglePlayback() {
        if isPlaying { stop() } else { start() }
    }

    func tapTempo() {
        let now = Date()
        tapTimestamps.append(now)
        if tapTimestamps.count > 4 {
            tapTimestamps.removeFirst(tapTimestamps.count - 4)
        }
        guard tapTimestamps.count >= 2 else { return }
        var totalInterval: TimeInterval = 0
        for i in 1..<tapTimestamps.count {
            totalInterval += tapTimestamps[i].timeIntervalSince(tapTimestamps[i - 1])
        }
        let avgInterval = totalInterval / Double(tapTimestamps.count - 1)
        if avgInterval > 0 {
            setBpm(60.0 / avgInterval)
        }
    }

    func start() {
        isPlaying = true
        measureCount = 0
        currentBeat = -1
        currentSubBeat = 0
        startTimer()
    }

    func stop() {
        timer?.cancel()
        timer = nil
        isPlaying = false
        currentBeat = -1
        currentSubBeat = 0
    }

    private func startTimer() {
        let interval = 60.0 / (bpm * Double(subdivision))
        var totalSubBeat = 0

        timer = Timer.publish(every: interval, on: .main, in: .common)
            .autoconnect()
            .sink { [weak self] _ in
                guard let self = self else { return }
                let beat = totalSubBeat / self.subdivision
                let subBeat = totalSubBeat % self.subdivision

                let wrappedBeat = beat % self.beatsPerMeasure
                self.currentBeat = wrappedBeat
                self.currentSubBeat = subBeat

                if wrappedBeat == 0 && subBeat == 0 && totalSubBeat > 0 {
                    self.measureCount += 1
                }

                let type = self.accentPattern.indices.contains(wrappedBeat)
                    ? self.accentPattern[wrappedBeat] : .normal

                switch type {
                case .mute:
                    break
                case .accent where subBeat == 0:
                    self.tonePlayer.play("click_high")
                case .normal where subBeat == 0:
                    self.tonePlayer.play("click_low")
                default:
                    if subBeat > 0 {
                        self.tonePlayer.play("click_low")
                    }
                }

                totalSubBeat += 1
            }
    }

    private func restartTimer() {
        timer?.cancel()
        startTimer()
    }

    private static func convertAccentPattern(_ kmpList: [shared.BeatType]) -> [BeatType] {
        kmpList.map { convertBeatType($0) }
    }

    private static func convertBeatType(_ kmp: shared.BeatType) -> BeatType {
        switch kmp {
        case .accent: return .accent
        case .normal: return .normal
        case .mute: return .mute
        default: return .normal
        }
    }

    private static func toKmpBeatType(_ local: BeatType) -> shared.BeatType {
        switch local {
        case .accent: return .accent
        case .normal: return .normal
        case .mute: return .mute
        }
    }
}
